### ISSUE TITLE
Clang's `-Wunused-template` false positive from requires clause

### ISSUE THREAD

**Issue by chandlerc** [llvm/llvm-project#218429]
`-Wunused-template` was recently added to `-Wunused`, but it seems to have a false positive.

https://cpp.compiler-explorer.com/z/e4b1Gr7jn

```c++
static void O(char /*arg*/) {}
static void O(int /*arg*/) {}

template <typename T>
static void A(T /*arg*/) {}

template <typename T>
static constexpr bool BT = requires {
  A<void(*)(T)>(O);
};

int main() {
  if constexpr (BT<int>) {
    O('x');
    O(42);
  }
}
```

Produces:

```
<source>:5:13: warning: unused function template 'A' [-Wunused-template]
    5 | static void A(T /*arg*/) {}
      |             ^
1 warning generated.
```

While *technically* the require clause will compile even without the template, it still uses the template to change its result.

**Comment by shafik** [comment]
Looks like [flash1729](https://github.com/flash1729) brought this in here https://github.com/llvm/llvm-project/pull/208001

It looks like there was some comments that this had false positives, maybe this should be reverted?

CC @cor3ntin

**Comment by cor3ntin** [comment]
I did not notice how disruptive the original PR was.
I think it makes sense to revert, certainly for 24.
